### DISCUSSION TITLE
v3.0 what's the meaning of the new method Language.resume_training ?

### DISCUSSION THREAD

**Discussion by rsoika** [explosion/spaCy#6902] [Help: Coding & Implementations] [answered]
I have a system running in a supervised training mode. Only the first time, when no model still exists I call the method nlp.initialize().

    if modelExists:
        nlp = spacy.load(modelPath) 
        # optimizer = nlp.resume_training()  # should I call this?
    else:
        nlp = spacy.blank(language) 
        optimizer=nlp.initialize()

     ....
     losses=nlp.update(examples, sgd=optimizer);

If I have alredy a model I start immediately with updating my model.

Now I have read about the methods `Language.resume_training` and `Language.rehearse` . Both methods are marked as '`experimental`' . From my tests so far, I can not see any changes in the results if I call these experimental methods or not. 

What is your recommendation in my scenario when I eventually have some new training data and I just want to update my existing model?

**Comment by svlandeg** [comment]
Hi! I'm going to convert your message to the discussion board, which is more suited for these kind of discussions among the community. We try to keep the issue tracker focused mainly on bugs.

You'll probably be getting an automatic email that this is closed but it really isn't - it'll just be converted and forwarded to the new thread ;-)

**Comment by polm** [comment, marked as answer]
Hey, sorry for the delayed reply on this. Maybe you saw this already but [this post](https://explosion.ai/blog/pseudo-rehearsal-catastrophic-forgetting) is the starting point of those methods, and should give you an overview on how we expect them to be used, if not specific examples. 

While the code for the methods is substantially complete, we haven't had a chance to test it properly or create good examples yet, so it's included in spaCy as something you can try out if you're feeling adventurous (and thus "experimental"). We hope to polish it up and document it properly but we don't have a fixed date planned.

> What is your recommendation in my scenario when I eventually have some new training data and I just want to update my existing model?

If you have all your training data, re-training from scratch will generally get the best performance, as all the training examples can be learned from in a balanced fashion. One of the reasons for rehearsal / resuming training is because we can't redistribute the training data for our pretrained models due to licensing issues.

**Comment by rsoika** [comment]
I don't know if my scenario is understandable.

I have a training model to detect entities in invoice documents. My software is using this model to support users in creating invoice approval requests. During this task the users refine the data and fix incorrect entities manually. For example: if SpaCy detects the wrong invoice date the user will fix it in our application for further processing.

As a result of this workflow my application is generating new training data all over the day. My idea is always resend a new single training data set of one invoice as soon as the users have confirmed all entities, so that SpaCy can refine its model. So I send one by one new training set over the time. Performance is not an issue here because this all is managed by a background process. 

Why should I retrain the complete set off invoices? Isn't is more efficient if I refine it over the time?
